testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/base/HttpClientTest.groovy

```diff
@@ -395,6 +404,49 @@ abstract class HttpClientTest extends AgentInstrumentationSpecification {
     method = "HEAD"
   }
 
+  /**
+   * This test fires a large number of concurrent requests.
+   * Each request first hits a HTTP server and then makes another client request.
+   * The goal of this test is to verify that in highly concurrent environment our instrumentations
+   * for http clients (especially inherently concurrent ones, such as Netty or Reactor) correctly
+   * propagate trace context.
+   */
+  def "causality test"() {
+    setup:
+    assumeTrue(testCausality())
+    int count = 50
+    def method = 'GET'
+    def frontUrl = server.address.resolve("/front")
+    def backUrl = server.address.resolve("/success")
+
+    def latch = new CountDownLatch(1)
+    def job = {
+      latch.await()
+      doRequest(method, frontUrl)
+    }
+
+    def pool = Executors.newFixedThreadPool(4)
+
+    when:
+    count.times {
+      pool.submit(job)
+    }
+    latch.countDown()
+
+    then:
+    assertTraces(count) {
+      count.times { traceIndex ->
+        trace(traceIndex, 4) {
+          clientSpan(it, 0, null, method, frontUrl)
+          serverSpan(it, 1, span(0))
```

Comment:
I wonder if we should also have a common span attribute that varies with each requests (maybe via a tiny bit of added manual instrumentation pulling a value from a header?) so we can very that we're not just "getting lucky" and happen to end up with 4 lined up spans on each request?

Reply:
You mean like https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/4d59fc0c8198343fbb2bd29a02f197ae48324815/instrumentation/kotlinx-coroutines/javaagent/src/test/groovy/KotlinCoroutineInstrumentationTest.groovy#L200 ?

Reply:
Yeah, I skipped that in hurry. I will see if I can do that easily enough.

Reply:
Done